pkg/operator/yugabytedb/controller.go

```diff
@@ -0,0 +1,952 @@
+		logger.Errorf("invalid cluster spec: %+v", err)
+		return
+	}
+
+	if err := c.createMasterHeadlessService(cluster); err != nil {
+		logger.Errorf("failed to create master headless service: %+v", err)
+		return
+	}
+
+	if err := c.createTServerHeadlessService(cluster); err != nil {
+		logger.Errorf("failed to create TServer headless service: %+v", err)
+		return
+	}
+
+	if err := c.createMasterUIService(cluster); err != nil {
+		logger.Errorf("failed to create Master UI service: %+v", err)
+		return
+	}
+
+	if err := c.createTServerUIService(cluster); err != nil {
+		logger.Errorf("failed to create replica service: %+v", err)
+		return
+	}
+
+	if err := c.createMasterStatefulset(cluster); err != nil {
+		logger.Errorf("failed to create master stateful set: %+v", err)
+		return
+	}
+
+	if err := c.createTServerStatefulset(cluster); err != nil {
+		logger.Errorf("failed to create tserver stateful set: %+v", err)
+		return
+	}
+
+	logger.Infof("succeeded creating and initializing cluster in namespace %s", cluster.namespace)
+}
+
+func (c *ClusterController) onUpdate(oldObj, newObj interface{}) {
+	_ = oldObj.(*yugabytedbv1alpha1.YBCluster).DeepCopy()
+	newObjCluster := newObj.(*yugabytedbv1alpha1.YBCluster).DeepCopy()
+	newYBCluster := newCluster(newObjCluster, c.context)
+
+	// Validate new spec
+	if err := validateClusterSpec(newYBCluster.spec); err != nil {
+		logger.Errorf("invalid cluster spec: %+v", err)
+		return
+	}
+
+	// Update headless service ports
+	if err := c.updateMasterHeadlessService(newYBCluster); err != nil {
+		logger.Errorf("failed to update Master headless service: %+v", err)
+		return
+	}
+
+	if err := c.updateTServerHeadlessService(newYBCluster); err != nil {
+		logger.Errorf("failed to update TServer headless service: %+v", err)
+		return
+	}
+
+	// Create/update/delete UI services (create/delete would apply for TServer UI services)
+	if err := c.updateMasterUIService(newYBCluster); err != nil {
+		logger.Errorf("failed to update Master UI service: %+v", err)
+		return
+	}
+
+	if err := c.updateTServerUIService(newYBCluster); err != nil {
+		logger.Errorf("failed to update TServer UI service: %+v", err)
+		return
+	}
+
+	// Update StatefulSets replica count, command, ports & PVCs.
+	if err := c.updateMasterStatefulset(newYBCluster); err != nil {
+		logger.Errorf("failed to update Master statefulsets: %+v", err)
+		return
+	}
+
+	if err := c.updateTServerStatefulset(newYBCluster); err != nil {
+		logger.Errorf("failed to update TServer statefulsets: %+v", err)
+		return
+	}
+
+	logger.Infof("cluster %s updated in namespace %s", newObjCluster.Name, newObjCluster.Namespace)
+}
+
+func (c *ClusterController) onDelete(obj interface{}) {
+	cluster, ok := obj.(*yugabytedbv1alpha1.YBCluster)
+	if !ok {
+		return
+	}
+	cluster = cluster.DeepCopy()
+	logger.Infof("cluster %s deleted from namespace %s", cluster.Name, cluster.Namespace)
+}
+
+func (c *ClusterController) createMasterUIService(cluster *cluster) error {
```

Comment:
This is a really long file. I'd recommend at least splitting up into separate .go files for tserver and other services.  New golang packages could even be useful, though not necessary.